system/include/emscripten/val.h

```diff
@@ -496,10 +496,52 @@ namespace emscripten {
     template<typename T>
     std::vector<T> vecFromJSArray(val v) {
         auto l = v["length"].as<unsigned>();
-
         std::vector<T> rv;
-        for(unsigned i = 0; i < l; ++i) {
-            rv.push_back(v[i].as<T>());
+
+        if (internal::typeSupportsMemoryView<T>()) {
+            rv.resize(l);
+
+            const char *arrayType;
+
+            switch (sizeof(T)) {
+                case 1:
+                    if (std::is_unsigned<T>::value) {
+                        arrayType = "Uint8Array";
+                    } else {
+                        arrayType = "Int8Array";
+                    }
+                    break;
+                case 2:
+                    if (std::is_unsigned<T>::value) {
+                        arrayType = "Uint16Array";
+                    } else {
+                        arrayType = "Int16Array";
+                    }
+                    break;
+                case 4:
+                    if (std::is_unsigned<T>::value) {
+                        arrayType = "Uint32Array";
+                    } else if (std::is_integral<T>::value) {
+                        arrayType = "Int32Array";
+                    } else {
+                        arrayType = "Float32Array";
+                    }
+                    break;
+                case 8:
+                    arrayType = "Float64Array";
+                    break;
+                default:
+                    throw;
+            }
+
+            val memory = val::module_property("buffer");
+            val memoryView = val::global(arrayType).new_(memory, reinterpret_cast<std::uintptr_t>(rv.data()), l);
```

Comment:
Where does the deletion of this `new_` occur?

Reply:
`new_` creates a JS `TypedArray` object, which cannot be explicitly deleted, but eventually will be collected by the JS engine Garbage Collector.

Reply:
Ah right, `val` is refcounted, and deleted once references drop to zero. I was thinking about the Embind objects that need to be manually `.delete()`d here.